### PR TITLE
JAMES-2555 ReIndexing using messageId with WebAdmin

### PR DESCRIPTION
**Pull request by chibenwa**
Solves https://github.com/linagora/james-project/pull/1802#discussion_r224726971

### CODE REVIEW

**Comment by chibenwa** [comment]
```

[9504bb26cafc98f692627703c0e63ce3256bc4f7] [ERROR]   RabbitMQClusterTest$ClusterSharing.queuesShouldBeDeclarableOnAnotherNode » Completion
[9504bb26cafc98f692627703c0e63ce3256bc4f7] [ERROR]   RabbitMQClusterTest$ClusterSharing.queuesShouldBeShared » Completion org.testc...
[9504bb26cafc98f692627703c0e63ce3256bc4f7] [ERROR]   RabbitMQClusterTest$ClusterSharing.rabbitMQManagerShouldReturnThreeNodesWhenAskingForStatus » Completion
```

test this please

File: server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/ReindexingRoutesTest.java [hunk 1]
```diff
@@ -705,4 +711,132 @@ void mailboxReprocessingShouldPerformReprocessingWhenMail() throws Exception {
...
+                    .post("/index/message/7?task=bad")
+                .then()
+                    .statusCode(HttpStatus.BAD_REQUEST_400)
+                    .body("statusCode", is(400))
+                    .body("type", is(ErrorResponder.ErrorType.INVALID_ARGUMENT.getType()))
+                    .body("message", is("task query parameter is mandatory. The only supported value is `reIndex`"));
+            }
+
+            @Test
+            void messageReprocessingShouldFailWithBadMessageId() {
```
Comment (rouazana) [inline server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/ReindexingRoutesTest.java]: s/message/messageId/

**Comment by chibenwa** [comment]
How about `/searchIndex` to avoid the name clash?

File: mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/MessageIdReIndexerImpl.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        this.mailboxSessionMapperFactory = mailboxSessionMapperFactory;
+        this.index = index;
+    }
+
+    @Override
+    public Task reIndex(MessageId messageId) {
+        return new MessageIdReIndexingTask(mailboxManager, mailboxSessionMapperFactory, index, messageId);
+    }
+
+    public static class MessageIdReIndexingTask implements Task {
```
Comment (trantienduchn) [inline mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/MessageIdReIndexerImpl.java]: static class goes first? right?

File: src/site/markdown/server/manage-webadmin.md [hunk 3]
```diff
@@ -2234,6 +2235,45 @@ Warning: During the re-indexing, the result of search operations might be altere
...
+Response codes:
+
+ - 201: Success. Corresponding task id is returned.
+ - 400: Error in the request. Details can be found in the reported error.
+
+The scheduled task will have the following type `MessageIdReIndexingTask` and the following `additionalInformation`:
+
+```
+{
+  "messageId":"18"
```
Comment (trantienduchn) [inline src/site/markdown/server/manage-webadmin.md]: I prefer cassandra UUID than a number MessageId, atleast it shows the slightly difference with `uid`

File: server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/MessageIdReindexingRoutes.java [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+import io.swagger.annotations.ApiResponse;
+import io.swagger.annotations.ApiResponses;
+import spark.Request;
+import spark.Response;
+import spark.Service;
+
+@Api(tags = "MessageIdReIndexing")
+@Path("/index")
+@Produces("application/json")
+public class MessageIdReindexingRoutes implements Routes {
```
Comment (trantienduchn) [inline server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/MessageIdReindexingRoutes.java]: why don't you place this route inside `ReIndexingRoutes` while placing test for this insde `ReIndexingRoutesTest`

File: mailbox/api/src/main/java/org/apache/james/mailbox/indexer/MessageIdReIndexer.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.mailbox.indexer;
+
+import org.apache.james.mailbox.model.MessageId;
+import org.apache.james.task.Task;
+
+public interface MessageIdReIndexer {
+    Task reIndex(MessageId messageId);
```
Comment (trantienduchn) [inline mailbox/api/src/main/java/org/apache/james/mailbox/indexer/MessageIdReIndexer.java]: why don't you place this method inside `ReIndexer` while creating a new interface for it?

File: src/site/markdown/server/manage-webadmin.md (hunk 3 above)
Comment (chibenwa) [reply src/site/markdown/server/manage-webadmin.md]: Well that depend on mailbox implementation...

File: mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/MessageIdReIndexerImpl.java (hunk 2 above)
Comment (chibenwa) [reply mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/MessageIdReIndexerImpl.java]: No visibility then static then filan then your stuff

File: server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/MessageIdReindexingRoutes.java (hunk 4 above)
Comment (chibenwa) [reply server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/MessageIdReindexingRoutes.java]: Well :trollface: I can't

You see, these routes can't be supported for JPA as JPA do not support messageId.

However it makes sens to have 'non-messageId' reIndexing routes for JPA.

Thus it makes sens to separate the routes to let the implementations bind the one they can support and the ones that makes sens.

File: mailbox/api/src/main/java/org/apache/james/mailbox/indexer/MessageIdReIndexer.java (hunk 5 above)
Comment (chibenwa) [reply mailbox/api/src/main/java/org/apache/james/mailbox/indexer/MessageIdReIndexer.java]: For the exact same reason described here https://github.com/linagora/james-project/pull/1808#discussion_r225387910

**Comment by chibenwa** [comment]
I ended up choosing `/mailboxIndex/...`

File: server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/MessageIdReindexingRoutes.java [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+        return BASE_PATH;
+    }
+
+    @Override
+    public void define(Service service) {
+        service.post(MESSAGE_PATH, this::reIndexMessage, jsonTransformer);
+    }
+
+    @POST
+    @Path("/message/{messageId}")
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/MessageIdReindexingRoutes.java]: `message` should be plural

**Comment by chibenwa** [comment]
```
[e8690121c91c9c38aa1ef8f27f1e09a2af97c50e] [ERROR]   SpamAssassinContract.movingAMailToTrashShouldNotImpactSpamassassinLearning:356 » ConditionTimeout
```

test this please

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**